Root init

jaysim opened the pull request:
1) Solve printing 'y/N' without any statement on interactive mode.

2) When initialize root index entries cause more than one of root entries are corrupted, do not delete system files and 'lost+found' directory for safety

3) Add progress bar and '-q' option for hiding progress bar

jaysim committed:
ntfsck: fix printing y/N without any output string in interactive mode

When execute ntfsck without any option, it will be execute with
interactive mode.
But at that time, show 'y/N' without any output string.
Fix it.

remove _ntfsck_ask_repair()

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: do not delete system files when initiaiize root directory

When root entries are corrupted, root directory index also should be
initialized. At that time, system files also removed.
And those system files added on phase of orphaned file checking.

But if fsck canceled or interrupted before adding system files, next
fsck does not work correctly.
So when root directory index is initialized, do not delete important
system files ($MFT/$MFTMirr ... $Extent) and lost+found directory

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
utils: make progress bar functions common

ntfsclone/ntfsresize/ntfscmp utils implemented and used
progress bar functions separately.

Make structures and functions related with progress bar common.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: add progress bar and quiet option for hiding progess

Add progress bar and quiet option for hiding progress bar.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: add structure/functions to organize fsck errors

Add header problem.h/problem.c which have structure to describe problem(error),
and add some function to check fsck option for repairing.

If '-p', '-a' will set PREEN option, it will hide some unnecessary print
statement. On the contrary, option '-y', '-n' will show print statement
with 'yes', or 'no'.

And also supports a structure that determine output(print) statements and
recovery depending on problem description.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

namjaejeon commented on libntfs-3g/volume.c:
```diff
@@ -2376,18 +2378,11 @@ BOOL _ntfsck_ask_repair(const ntfs_volume *vol, BOOL flag)
 					return TRUE;
 				else if (strcasecmp(answer, "\n") == 0 ||
-					 strcasecmp(answer, "N\n") == 0)
+						strcasecmp(answer, "N\n") == 0)
```
if we click the enter key, what happen ? only print "(y/N)?"

jaysim replied:
I just moved code ntfs_ask_repair() to _ntfs_ask_repair(),
hmm I think if we click only enter key, ntfsck reprint "(y/N)?" and ask your input again.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1180,11 +1188,10 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
 
-		fsck_err_found();
-		ntfs_log_debug("Clear the bit of mft no(%"PRId64") "
+		ntfs_log_error("Clear the bit of mft no(%"PRId64") "
```
we don't need to call fsck_err_found() here.

jaysim replied:
this checked with orphaned inode cadidiates which is all inodes except already checked inode in tree traverse. So, I confuse it is error or not.
I already modifed that fsck set bitmaps of all inodes which can be opened at first step temporarily. So, clearing bitmap without checking if those bitmap is set or not, is error ???
because this mismatch can be caused from temporary bitmap set at first step, I decide to remove err_found()

namjaejeon commented on src/ntfsck.c:
```diff
@@ -382,10 +383,16 @@ static int ntfsck_check_runlist(ntfs_attr *na, u8 set_bit, struct rl_size *rls,
 	}
 
 	if (dup_rl) {
-		check_failed("Found cluster duplication of inode(%"PRId64":%d), Fix it\n",
```
you replace check_failed() with ntfs_log_error(). If so, fsck_err_found() is not called. no problem ?

jaysim replied:
there is some consideration about this point.

check_failed() is called outside of this function(ntfsck_check_runlist())
because ntfsck_check_attr_runlist() check two kinds of corruption which runlist corruption and cluster duplication in ntfsck_decompose_runlist() and ntfsck_check_runlist() respectively, but applying repaired cluster run to disk is one place outside of ntfsck_check_attr_runlist().

If we count error inside function, error may be counted up to 2. but just error fixing occurred one place, it due to decrease just one error count. So, even though fsck fixed all error, fsck misundertood error is remained

I don't like this style of implementation. if you have any idea, let me know.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1944,6 +1944,180 @@ static runlist *ntfsck_decompose_runlist(ntfs_attr *na, BOOL *need_fix)
+		goto out;
+
+	/* remain one index block not to allocate when add index for meta files in fsck */
+	if (ntfs_attr_truncate(ia_na, block_size))
```
we open system files and add index entry for them in ntfsck_initiaiize_root_index(). Is there any reason why one index block should be remained ? because ntfs_index_add_filename will create index block if there is no room in index root ?

jaysim replied:
the reason is to remove possibility of duplicated cluster allocation in fsck.
to remove possibility, remain one $INDEX_ALLOCATION block whose entries length is initialized.
Then ntfs_index_add_filename will not allocate during add meta system files indexes.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -1163,8 +1171,10 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
 	u64 parent_no;
 	int nlink;
 
+#if 0
```
any reason why this should commented ?

jaysim replied:
I'll remove it, it moved outside of function.

namjaejeon commented on src/ntfsck.c:
```diff
@@ -4413,12 +4458,12 @@ int main(int argc, char **argv)
 	ntfs_log_set_levels(NTFS_LOG_LEVEL_INFO);
 	ntfs_log_clear_levels(NTFS_LOG_LEVEL_TRACE|NTFS_LOG_LEVEL_ENTER|NTFS_LOG_LEVEL_LEAVE);
-
+	pb_flags = NTFS_PROGBAR;
```
where is this used ?

jaysim replied:
if someone who do not want print 'progress bar', he can use option '-q'.
it set pb_flag |= ~NTFS_PROGBAR, then it set prgress_init() with no progress bar option.

jaysim committed:
ntfsck: apply fsck problem structure

Now, define structure related with fsck problem, apply it to all codes
which related with fsck repairing and printing statement, .

You may set problem_context_t structure for print statement, And call
ntfs_fix_problem with error code (repair code).
Then problem structure will print proper statement and parameters.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
doc: add/modify documents

Add README.md and ntfsck.8.in/ntfsprogs-plus.8.in for man page
Add Changelog and modify some docuements

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
all : add license statement to files

Add license statment on top of files like below.
/* SPDX-License-Identifier : GPL-2.0 */

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
all: substitute ntfs-3g to ntfs in all sources.

Rename "libntfs-3g" to "libntfs", and change all script for it.
Remove "include/ntfs-3g", and move all header files to "include/" itself.
Modify configure.ac and Makefile.am to remove ntfs-3g.
Remove ntfs-3g in utilities' version info.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

namjaejeon commented:
Applied it to #master now.
Thanks!
